spec/javascripts/app/document-capture/components/document-capture-spec.jsx

```diff
@@ -1,11 +1,8 @@
 import React from 'react';
 import { render } from '@testing-library/react';
 import DocumentCapture from '../../../../../app/javascript/app/document-capture/components/document-capture';
-import { useDOM } from '../../../support/dom';
 
 describe('document-capture/components/document-capture', () => {
-  useDOM();
```

Comment:
Do you think it would be worth splitting up the test suite for React things vs legacy things? I like how explicit these were

Reply:
The primary challenge this pull request seeks to address is that many dependencies will reach to assumed DOM globals as a side effect of merely being brought into scope (imported).

Examples:

- [`classList.js`](https://github.com/eligrey/classList.js/blob/ecb3305730a9b62f31cfdc71190d2f312233ed32/classList.js#L14)
- [`tabbable`, via `focus-trap`](https://github.com/davidtheclark/tabbable/blob/master/src/index.js#L14-L19)
- [`react-testing-library`](https://github.com/testing-library/dom-testing-library/blob/5b2640aca66b9ccf611baa88188e0807df3a73c7/src/helpers.js#L1) (a new discovery)

So if the DOM setup is delayed until this point in the test lifecycle, it's too late, since those have already been evaluated.

This is true for both legacy code and for the new React code, at least if it brings in those dependencies (like proposed with Accordion usage in #3938).

Putting the DOM setup in `spec_helper.js` helps ensure those globals are assigned in time for that initial import.

It's true that this means all tests will run with a DOM available, which isn't always necessary, and could potentially encourage (or at least not discourage) best practices with tight coupling to the DOM.

Because we'll need the DOM for the React tests, and the DOM will need to be set up early, I'm not sure it's something we could easily split, at least not in a way where we have some expressiveness like this within the test itself to declare it as needing the DOM.

One possible option could be to invert the behavior:

1. Always set up the DOM globals in the test setup so they're available at the time that dependencies are imported
2. _Unset_ those DOM globals in global test lifecycle methods
3. Allow an explicit opt-in (like this `useDOM`) which restores the DOM globals

I could explore this if it seems like a valuable addition.

That said, my experience is that the DOM will be needed more often than not, and it may not be worth the effort or the overhead in marking each of these tests explicitly.

Reply:
That's a good enough reason, thanks for explaining!

Reply:
As a general pattern though, I do think it could be nice to adopt this sort of convention of lifecycle-based opt-in helpers ("hooks").

Related: https://github.com/WordPress/gutenberg/issues/22804